File: apps/mobile_flutter/lib/core/api/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/app_config.dart';

class ApiClient {
  const ApiClient(this.config);

  final AppConfig config;

  Uri endpoint(String path) {
    final normalizedPath = path.startsWith('/') ? path : '/$path';
    return Uri.parse('${config.apiBaseUrl}$normalizedPath');
  }

  Future<List<dynamic>> getList(String path) async {
    final response = await http.get(endpoint(path));
    _throwIfUnsuccessful(response);
    return jsonDecode(response.body) as List<dynamic>;
  }

  Future<Map<String, dynamic>> getMap(String path) async {
    final response = await http.get(endpoint(path));
    _throwIfUnsuccessful(response);
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> postMap(
    String path,
    Map<String, dynamic> payload,
  ) async {
    final response = await http.post(
      endpoint(path),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(payload),
    );
    _throwIfUnsuccessful(response);
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  void _throwIfUnsuccessful(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ApiException(
        statusCode: response.statusCode,
        body: response.body,
      );
    }
  }
}

class ApiException implements Exception {
  const ApiException({
    required this.statusCode,
    required this.body,
  });

  final int statusCode;
  final String body;

  @override
  String toString() => 'ApiException($statusCode): $body';
}

File: apps/web_pwa/src/core/api/apiClient.ts
import type { AppConfig } from '../config/appConfig';

// Mirrors apps/mobile_flutter/lib/core/api/api_client.dart.
export class ApiException extends Error {
  readonly statusCode: number;
  readonly body: string;

  constructor(statusCode: number, body: string) {
    super(`ApiException(${statusCode}): ${body}`);
    this.statusCode = statusCode;
    this.body = body;
  }
}

export class ApiClient {
  private readonly config: AppConfig;

  constructor(config: AppConfig) {
    this.config = config;
  }

  private endpoint(path: string): string {
    const normalizedPath = path.startsWith('/') ? path : `/${path}`;
    return `${this.config.apiBaseUrl}${normalizedPath}`;
  }

  private async throwIfUnsuccessful(response: Response): Promise<void> {
    if (!response.ok) {
      const body = await response.text();
      throw new ApiException(response.status, body);
    }
  }

  async getList<T = unknown>(path: string): Promise<T[]> {
    const response = await fetch(this.endpoint(path));
    await this.throwIfUnsuccessful(response);
    return (await response.json()) as T[];
  }

  async getMap<T = Record<string, unknown>>(path: string): Promise<T> {
    const response = await fetch(this.endpoint(path));
    await this.throwIfUnsuccessful(response);
    return (await response.json()) as T;
  }

  async postMap<T = Record<string, unknown>>(
    path: string,
    payload: Record<string, unknown>,
  ): Promise<T> {
    const response = await fetch(this.endpoint(path), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    await this.throwIfUnsuccessful(response);
    return (await response.json()) as T;
  }
}
